crates/sui-core/src/authority/shared_object_version_manager.rs

```diff
@@ -0,0 +1,204 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::authority::epoch_start_configuration::EpochStartConfigTrait;
+use crate::authority::AuthorityPerEpochStore;
+use crate::execution_cache::ExecutionCacheRead;
+use std::collections::HashMap;
+use sui_types::crypto::RandomnessRound;
+use sui_types::effects::{TransactionEffects, TransactionEffectsAPI};
+use sui_types::executable_transaction::VerifiedExecutableTransaction;
+use sui_types::storage::{
+    transaction_input_object_keys, transaction_receiving_object_keys, ObjectKey,
+};
+use sui_types::transaction::{
+    SenderSignedData, SharedInputObject, TransactionDataAPI, TransactionKey,
+};
+use sui_types::{
+    base_types::{ObjectID, SequenceNumber},
+    error::SuiResult,
+    SUI_RANDOMNESS_STATE_OBJECT_ID,
+};
+use tracing::{debug, trace};
+
+pub struct SharedObjVerManager {}
+
+pub type AssignedTxAndVersions = Vec<(TransactionKey, Vec<(ObjectID, SequenceNumber)>)>;
+
+#[must_use]
+#[derive(Default)]
+pub struct ConsensusSharedObjVerAssignment {
+    pub shared_input_next_versions: HashMap<ObjectID, SequenceNumber>,
+    pub assigned_versions: AssignedTxAndVersions,
+}
+
+impl SharedObjVerManager {
+    pub async fn assign_versions_from_consensus(
+        epoch_store: &AuthorityPerEpochStore,
+        cache_reader: &dyn ExecutionCacheRead,
+        certificates: &[VerifiedExecutableTransaction],
+        randomness_round: Option<RandomnessRound>,
+    ) -> SuiResult<ConsensusSharedObjVerAssignment> {
+        let mut shared_input_next_versions = get_or_init_versions(
+            certificates.iter().map(|cert| cert.data()),
+            epoch_store,
+            cache_reader,
+            randomness_round.is_some(),
+        )
+        .await?;
+        let mut assigned_versions = Vec::new();
+        // We must update randomness object version first before processing any transaction,
+        // so that all reads are using the next version.
+        // TODO: Add a test that actually check this, i.e. if we change the order, some test should fail.
+        if let Some(round) = randomness_round {
+            // If we're generating randomness, update the randomness state object version.
+            let version = shared_input_next_versions
+                .get_mut(&SUI_RANDOMNESS_STATE_OBJECT_ID)
+                .expect("added randomness state object to lookup above");
```

Comment:
this is no longer "above" right? but in `get_or_init_versions`. (I guess the call to that function is "above" but it's a bit more confusing now bc the actual code that does it is below)